File: src/main/java/com/tms/csp/transforms/UntilStableTransformer.java
package com.tms.csp.transforms;

import com.tms.csp.ast.Exp;

import javax.annotation.Nonnull;

public class UntilStableTransformer extends Transformer {



    private Transformer transformer;

    public UntilStableTransformer(Transformer... transformer) {
        this(null, transformer);
    }

    public UntilStableTransformer(String name, Transformer... transformer) {
        this.name = name;
        if (transformer.length > 1) {
            this.transformer = new CompoundTransformer(transformer);
        } else {
            this.transformer = transformer[0];
        }
    }

    @Nonnull
    @Override
    public Exp transform(@Nonnull Exp in) {
        return transformUntilStable(transformer, in);
    }

}

File: src/main/java/com/tms/csp/transforms/UntilStableTransformer.kt
package com.tms.csp.transforms

import com.tms.csp.ast.Exp

class UntilStableTransformer(name: String?, vararg transformer: Transformer) : Transformer() {


    private var transformer: Transformer? = null

    constructor(vararg transformer: Transformer) : this(null, *transformer) {}

    init {

        this._name = name
        if (transformer.size > 1) {
            this.transformer = CompoundTransformer(*transformer)
        } else {
            this.transformer = transformer[0]
        }
    }

    override fun transform(inExp: Exp): Exp {
        return Transformer.Companion.transformUntilStable(transformer!!, inExp)
    }

}
